metrics/src/agent.rs

```diff
@@ -0,0 +1,267 @@
+                    Ok(metrics_str) => metrics_str
+                        .split('\n')
+                        .filter_map(|metric_str| Metric::from_string(metric_str, unix_timestamp))
+                        .collect(),
+                    Err(e) => {
+                        println!("Error converting metric to str: {}", e);
+                        continue;
+                    }
+                };
+
+                let mut metrics_to_process = buf_producer.lock().await;
+                for metric in metrics {
+                    metrics_to_process.push(metric);
+                }
+            }
+        });
+
+        // Process metrics we've parsed and flush them to Datadog
+        loop {
+            // We flush every three seconds so we don't send too many requests to Datadog
+            // in a short amount of time. It's a good default for now.
+            tokio::time::sleep(time::Duration::from_millis(3000)).await;
+
+            let mut metrics_to_process = buf_consumer.lock().await;
+
+            // We don't have any new metrics
+            if metrics_to_process.is_empty() {
+                continue;
+            }
+
+            let payload = match construct_distribution_payload(metrics_to_process.to_vec()) {
+                Ok(payload) => payload,
+                Err(e) => {
+                    println!("Error serializing payload: {}", e);
+                    continue;
+                }
+            };
+
+            println!("Sending payload: {}", payload);
+
+            // Create a POST request with the headers and payload
+            let request_option = Request::builder()
+                .method("POST")
+                .uri(format!(
+                    "https://api.{}/api/v1/distribution_points",
+                    self.config.site
+                ))
+                .header("DD-API-KEY", &self.config.api_key)
+                .header(CONTENT_TYPE, HeaderValue::from_static("application/json"))
+                .body(Body::from(payload.to_string()));
+
+            let request = match request_option {
+                Ok(request) => request,
+                Err(e) => {
+                    println!("Error constructing payload: {}", e);
+                    continue;
+                }
+            };
+
+            // Send the request and handle the response
+            match http_client.request(request).await {
+                Ok(_) => {
+                    println!("Successfully posted request");
+                    // Remove all elements from the buffer, as they've
+                    // been sent at this point
+                    metrics_to_process.clear();
+                }
+                Err(e) => {
+                    println!("Error sending request to Datadog: {}", e);
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Config {
+    api_key: String,
+    site: String,
+}
+
+pub struct ConfigBuilder {
```

Comment:
https://docs.rs/derive_builder/latest/derive_builder/ Is another alternative that doesn't add any runtime overhead.

Reply:
Oh that's awesome, I didn't know we could do it automagically! I'll see if I can get it working.

Reply:
Done, works like magic :-)